docs/proposals/server-side-apply.md

```diff
@@ -0,0 +1,233 @@
+---
+title: Server-Side Apply
+
+authors:
+- "@leoluz"
+
+sponsors:
+- TBD
+
+reviewers:
+- TBD
+
+approvers:
+- TBD
+
+creation-date: 2022-03-17
+
+---
+
+# Server-Side Apply support for ArgoCD
+
+[Server-Side Apply (SSA)][1] allows calculating the final patch to update
+resources in Kubernetes in the server instead of the client. This proposal
+describes how ArgoCD can leverage SSA during syncs.
+
+* [Open Questions](#open-questions)
+    * [[Q-1] How to handle conflicts?](#q-1-how-to-handle-conflicts)
+    * [[Q-2] Should we support multiple managers?](#q-2-should-we-support-multiple-managers)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Better interoperability with Admission Controllers](#better-interoperability-with-admission-controllers)
+    * [Better resource conflict management](#better-resource-conflict-management)
+    * [Better CRD support](#better-crd-support)
+* [Goals](#goals)
+    * [[G-1] Fine grained configuration](#g-1-fine-grained-configuration)
+    * [[G-2] Strategic merge patch while diffing](#g-2-strategic-merge-patch-while-diffing)
+    * [[G-3] Admission Controllers compatibility](#g-3-admission-controllers-compatibility)
+    * [[G-4] Conflict management](#g-4-conflict-management)
+    * [[G-5] Register a proper manager](#g-5-register-a-proper-manager)
+* [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Non-Functional Requirements](#non-functional-requirements)
+    * [Use cases](#use-cases)
+        * [[UC-1]: enable SSA at the controller level](#uc-1-as-a-user-i-would-like-enable-ssa-at-the-controller-level-so-all-application-are-applied-server-side)
+        * [[UC-2]: enable SSA at the Application level](#uc-2-as-a-user-i-would-like-enable-ssa-at-the-application-level-so-all-resources-are-applied-server-side)
+        * [[UC-3]: enable SSA at the resource level](#uc-3-as-a-user-i-would-like-enable-ssa-at-the-resource-level-so-only-a-single-manifest-is-applied-server-side)
+    * [Security Considerations](#security-considerations)
+    * [Risks and Mitigations](#risks-and-mitigations)
+        * [[R-1] Supported K8s version check](#r-1-supported-k8s-version-check)
+        * [[R-2] Alternating Server-Side Client-Side syncs](#r-2-alternating-server-side-client-side-syncs)
+    * [Upgrade / Downgrade](#upgrade--downgrade)
+* [Drawbacks](#drawbacks)
+
+---
+
+## Open Questions
+
+### [Q-1] How to handle conflicts?
+When SSA is enabled, the server may return field conflicts with other managers.
+What ArgoCD controller should do in case of conflict? Just force the sync and
+log warnings (like some other controllers do?)
+
+### [Q-2] Should we support multiple managers?
```

Comment:
I would not object to add an system level setting that allows users to override manager name.

Reply:
I suspect that this is a case where it could bring some strange corner cases. It might be simple to implement but maybe not that simple to deal with future bugs caused by it. 

Multi-manager feature at first seems odd to me as I haven't seen controllers using this pattern. In my experience, what I've seen is controllers defining a single manager identifier for them. 

If we don't have a clear (and necessary) use-case for it, I would vote to keep multi-manager out of this proposal.